File: Beetle.NetPackage/Samples/c#/BeetleNPDemo/BeetleNPDemo/Beetle.NetPackage/NetClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Sockets;
using System.Net;
/**
 * Copyright © henryfan 2013
 * Created by henryfan on 13-7-30.
 * homepage:www.ikende.com
 * email:henryfan@msn.com
 */
namespace Beetle.NetPackage
{
    public class NetClient : IDisposable
    {

        public NetClient(string host, int port, Package package, INetClientHandler handler)
        {


            mEndPoint = new DnsEndPoint(host, port);
            Handler = handler;
            mPackage = package;
            mPackage.Receive = OnDataReceive;
            mReceiveSAEA = new SocketAsyncEventArgs();
            mReceiveSAEA.Completed += ReceiveCompleted;
            mReceiveSAEA.SetBuffer(new byte[1024 * 4], 0, 1024 * 4);

            mConnectSAEA = new SocketAsyncEventArgs();
            mConnectSAEA.RemoteEndPoint = mEndPoint;
            mConnectSAEA.Completed += OnConnected;

            mSendSAEA = new SocketAsyncEventArgs();
            mSendSAEA.Completed += SendCompleted;

        }

        private System.Net.Sockets.SocketAsyncEventArgs mConnectSAEA;

        private System.Net.Sockets.SocketAsyncEventArgs mSendSAEA;

        private System.Net.Sockets.SocketAsyncEventArgs mReceiveSAEA;

        private Package mPackage;

        public INetClientHandler Handler
        {
            get;
            set;
        }

        private System.Net.EndPoint mEndPoint;

        private System.Net.Sockets.Socket mSocket;

        private bool mConnected = false;

        private bool mLittleEndian = true;

        public bool Connected
        {
            get
            {
                return mConnected;
            }
        }

        public Exception LastError
        {
            get;
            private set;
        }

        private void OnConnected(object sender, SocketAsyncEventArgs e)
        {
            if (e.SocketError == SocketError.Success)
            {
                mConnected = true;
                BeginReceive();
                Handler.Connected(this);
            }
            else
            {
                mConnected = false;
                ReleaseSocket();
            }
        }

        public bool Connect()
        {
            if (Connected)
                return true;
            try
            {
                mPackage.Reset();
                mSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                if (!mSocket.ConnectAsync(mConnectSAEA))
                {
                    OnConnected(this, mConnectSAEA);
                }
                return mConnected;
            }
            catch (Exception e_)
            {
                mConnected = false;
                OnError(e_);
                return false;
            }

        }

        private void OnError(Exception e)
        {
            try
            {
                Handler.ClientError(this, e);
            }
            catch
            {

            }
        }

        private void BeginReceive()
        {
            try
            {
                if (!mSocket.ReceiveAsync(mReceiveSAEA))
                {
                    ReceiveCompleted(this, mReceiveSAEA);
                }
            }
            catch (SocketException se)
            {
                ReleaseSocket();
                OnError(se);
            }
            catch (ObjectDisposedException ode)
            {
                ReleaseSocket();
                OnError(ode);
            }
            catch (Exception e_)
            {
                OnError(e_);

            }
        }

        private void ReceiveCompleted(object sender, System.Net.Sockets.SocketAsyncEventArgs e)
        {
            if (e.SocketError == SocketError.Success && e.BytesTransferred > 0)
            {
                try
                {
                    mPackage.Import(e.Buffer, 0, e.BytesTransferred);
                    BeginReceive();
                }
                catch (Exception e_)
                {
                    ReleaseSocket();
                    OnError(e_);
                }

            }
            else
            {
                ReleaseSocket();
            }

        }

        private void SendCompleted(object sender, System.Net.Sockets.SocketAsyncEventArgs e)
        {
            if (e.SocketError == SocketError.Success && e.BytesTransferred > 0)
            {
                if (e.BytesTransferred < e.Count)
                {
                    e.SetBuffer(e.Offset + e.BytesTransferred, e.Count - e.BytesTransferred);
                    if (!mSocket.SendAsync(e))
                    {
                        SendCompleted(this, e);
                    }
                    

                }
            }
            else
            {
                ReleaseSocket();
            }
        }

        private void OnDataReceive(System.IO.Stream stream)
        {
            try
            {
                stream.Position = 0;
                BufferStream reader = new BufferStream(stream, LittleEndian);
                IMessage msg = mPackage.FromStream(reader);
                if (msg == null)
                    throw new Exception("message type not found!");
                Handler.ClientReceive(this, mPackage.ReceiveCast(msg));
            }
            catch (Exception e_)
            {
                OnError(e_);
            }
            finally
            {
                mPackage.Reset();
            }
        }

        public bool LittleEndian
        {
            get
            {
                return mLittleEndian;
            }
            set
            {
                mLittleEndian = value;
                mPackage.LittleEndian = LittleEndian;
            }
        }

        public bool Send(object message)
        {
            lock (this)
            {
                if (Connect())
                {

                    try
                    {
                        IMessage msg = mPackage.SendCast(message);
                        byte[] data = mPackage.GetMessageData(msg);
                        mSendSAEA.SetBuffer(data, 0, data.Length);
                        if (!mSocket.SendAsync(mSendSAEA))
                        {
                            SendCompleted(this, mSendSAEA);
                        }
                        return true;
                    }
                    catch (SocketException se)
                    {
                        ReleaseSocket();
                        OnError(se);
                    }
                    catch (ObjectDisposedException ode)
                    {
                        ReleaseSocket();
                        OnError(ode);
                    }
                    catch (Exception e_)
                    {
                        OnError(e_);

                    }
                }
                return false;
            }

        }

        private void ReleaseSocket()
        {
            mConnected = false;
            lock (this)
            {
                if (mSocket != null)
                {
                    Handler.ClientDisposed(this);
                    try
                    {
                        if (mSocket != null)
                        {
                            mSocket.Shutdown(SocketShutdown.Both);
                        }
                    }
                    catch
                    {
                    }
                    try
                    {
                        if (mSocket != null)
                        {
                            mSocket.Close();
                        }
                    }
                    catch
                    {
                    }
                }
                mSocket = null;

            }

        }

        public void Dispose()
        {
            ReleaseSocket();
        }
    }
}

File: Beetle.NetPackage/Sources/Cpp/NetClient.cpp
#include "Stdafx.h"
#include "INetClientHandler.h"
#include "NetClient.h"
#include "Package.h"
#include "BufferStream.h"

using namespace System::Threading;

namespace Beetle
{
	namespace NetPackage
	{
		NetClient::NetClient( String ^host, int port, Package ^package, INetClientHandler ^handler )
		{
			m_Connected = false;
			m_LittleEndian =true;
			array<System::Net::IPAddress ^> ^ips = System::Net::Dns::GetHostAddresses(host);
			m_Host = ips[0];
			m_Port = port;
			Handler = handler;
			m_Package = package;
			m_Package->Receive = gcnew Action<System::IO::Stream ^>(this, &NetClient::OnDataReceive);
			m_ReceiveSAEA = gcnew SocketAsyncEventArgs();
			m_ReceiveSAEA->Completed += gcnew EventHandler<SocketAsyncEventArgs ^>(this, &NetClient::ReceiveCompleted);
			m_ReceiveSAEA->SetBuffer(gcnew array<Byte>(1024 * 4), 0, 1024 * 4);
		}

		NetClient::~NetClient()
		{
			ReleaseSocket();
		}

		bool NetClient::LittleEndian::get(void) 
		{ 
			return m_LittleEndian; 
		}
		void NetClient::LittleEndian::set(bool val) 
		{
			m_LittleEndian = val;
			m_Package->LittleEndian = LittleEndian;
		}

		bool NetClient::Connect()
		{
			if (Connected)
			{
				return true;
			}
			try
			{
				m_Package->Reset();
				m_Socket = gcnew Socket(AddressFamily::InterNetwork, SocketType::Stream, ProtocolType::Tcp);
				m_Socket->Connect(m_Host, m_Port);
				BeginReceive();
				m_Connected = true;
				Handler->Connected(this);
				return true;
			}
			catch (Exception ^e_)
			{
				m_Connected = false;
				OnError(e_);
				return false;
			}
		}

		void NetClient::OnError( Exception ^e )
		{
			try
			{
				Handler->ClientError(this, e);
			}
			catch (Exception ^e)
			{
				throw gcnew Exception(e->Message);
			}
		}

		void NetClient::BeginReceive()
		{
			try
			{
				if (!m_Socket->ReceiveAsync(m_ReceiveSAEA))
				{
					ReceiveCompleted(this, m_ReceiveSAEA);
				}
			}
			catch (SocketException ^se)
			{
				ReleaseSocket();
				OnError(se);
			}
			catch (ObjectDisposedException ^ode)
			{
				ReleaseSocket();
				OnError(ode);
			}
			catch (Exception ^e_)
			{
				OnError(e_);
			}
		}

		void NetClient::ReceiveCompleted( Object ^sender, System::Net::Sockets::SocketAsyncEventArgs ^e )
		{
			if (e->SocketError == SocketError::Success && e->BytesTransferred > 0)
			{
				try
				{
					m_Package->Import(e->Buffer, 0, e->BytesTransferred);
					BeginReceive();
				}
				catch (Exception ^e_)
				{
					ReleaseSocket();
					OnError(e_);
				}
			}
			else
			{
				ReleaseSocket();
			}
		}

		void NetClient::OnDataReceive( System::IO::Stream ^stream )
		{
			try
			{
				stream->Position = 0;
				BufferStream ^reader = gcnew BufferStream(stream, LittleEndian);
				IMessage ^msg = m_Package->FromStream(reader);
				if (msg == nullptr)
				{
					throw gcnew Exception(L"message type not found!");
				}
				Handler->ClientReceive(this, m_Package->DoReceiveCast(msg));
			}
			catch (Exception ^e_)
			{
				OnError(e_);
			}
			finally
			{
				m_Package->Reset();
			}
		}

		bool NetClient::Send( Object ^message )
		{
			bool flag = false;
			Monitor::Enter(this);
			if (Connect())
			{
				try
				{
					IMessage ^msg = m_Package->DoSendCast(message);
					array<Byte> ^data = m_Package->GetMessageData(msg);
					int count = data->Length;
					int index = 0;
					int scount;
					while (count > 0)
					{
						scount = m_Socket->Send(data, index, count, SocketFlags::None);
						count -= scount;
						index += scount;
					}
					flag = true;
				}
				catch (SocketException ^se)
				{
					ReleaseSocket();
					OnError(se);
				}
				catch (ObjectDisposedException ^ode)
				{
					ReleaseSocket();
					OnError(ode);
				}
				catch (Exception ^e_)
				{
					OnError(e_);
				}
			}
			Monitor::Exit(this);
			return flag;
		}

		void NetClient::ReleaseSocket()
		{
			m_Connected = false;
			Monitor::Enter(this);
			if (m_Socket != nullptr)
			{
				Handler->ClientDisposed(this);
				try
				{
					if (m_Socket != nullptr)
					{
						m_Socket->Shutdown(SocketShutdown::Both);
					}
				}
				catch (Exception ^e)
				{
					throw gcnew Exception(e->Message);
				}
				try
				{
					if (m_Socket != nullptr)
					{
						m_Socket->Close();
					}
				}
				catch (Exception ^e)
				{
					throw gcnew Exception(e->Message);
				}
			}
			m_Socket = nullptr;
			Monitor::Exit(this);
		}

	}
}